README
Utilities to convert metadata for upgrade from elliptics 2.9 to 2.10

There is 3 steps to convert metadata from 2.9 to 2.10:

1. Convert Kyoto Cabinet meta.kch to eblob meta using dnet_convert_meta utility
	dnet_convert_meta -M /path/to/meta.kch -N /path/to/eblob-meta
   This step is mandatory.

2. Convert Kyoto Cabinet history.kch to create META_UPDATE timestamps that are required for correct checks
	dnet_convert_history -M /path/to/eblob-meta -H /path/to/history.kch -g 1:2
   If there is records in history.kch that doesn't exists in meta.kch this utility will create it. -g specifies groups for such records.
   This utility is not mandatory but it's highly recommended to run it.

3. Run over files on filesystem/eblob to add missed meta records and optionally update checksums
	dnet_convert_files --input-path /path/to/files/root --meta /path/to/eblob-meta --group 1 --group 2 
   Input path should point to root directory in case of filesystem backand or to eblob in case of eblob backend.
   If there is files that doesn't have records in meta thils utility will create it. --group parameter specifies groups for such records.
   There are optional parameters:
     --threads (default it 16) - number of threads that iterates over eblob/filesystem
     --enable-checksum - enable checksum calculation and update. If old checksum differs this utility will overwrite it.



debian/changelog
elliptics-convert-meta (0.2.8) lucid; urgency=low

  * Updated to use eblob >= 0.12

 -- Anton Kortunov <toshic@yandex-team.ru>  Mon, 07 Nov 2011 19:14:56 +0300

elliptics-convert-meta (0.2.7) lucid; urgency=low

  * Added --update-date option

 -- Anton Kortunov <toshik@yandex-team.ru>  Fri, 22 Aug 2011 21:56:02 +0400

elliptics-convert-meta (0.2.6) lucid; urgency=low

  * Correct shared pointers usage in eblob_processor

 -- Anton Kortunov <toshik@yandex-team.ru>  Fri, 05 Aug 2011 10:20:02 +0400

elliptics-convert-meta (0.2.5) lucid; urgency=low

  * Return correct error code
  * Remove Crupto++ include

 -- Anton Kortunov <toshik@yandex-team.ru>  Thu, 04 Aug 2011 13:38:39 +0400

elliptics-convert-meta (0.2.4) unstable; urgency=low

  * Changed elliptics package name to elliptics-2.10

 -- Anton Kortunov <toshik@yandex-team.ru>  Mon, 01 August 2011 17:50:00 +0300

elliptics-convert-meta (0.2.3) unstable; urgency=low

  * Fixed checksum calculation
  * Added blob_unsort utility (converts blob.0.index.sorted to blob.0.index)

 -- Anton Kortunov <toshik@yandex-team.ru>  Mon, 01 August 2011 16:30:00 +0300

elliptics-convert-meta (0.2.2) unstable; urgency=low

  * Fixed Debian rules file

 -- Anton Kortunov <toshik@yandex-team.ru>  Mon, 25 July 2011 17:30:00 +0300

elliptics-convert-meta (0.2.1) unstable; urgency=low

  * Fixed Linux Hardy build

 -- Anton Kortunov <toshik@yandex-team.ru>  Mon, 25 July 2011 17:30:00 +0300

elliptics-convert-meta (0.2.0) unstable; urgency=low

  * Added checksum support

 -- Anton Kortunov <toshik@yandex-team.ru>  Mon, 25 July 2011 17:00:00 +0300

elliptics-convert-meta (0.1.0) unstable; urgency=low

  * Initial release

 -- Anton Kortunov <toshik@yandex-team.ru>  Fri, 15 July 2011 18:00:00 +0300


common.c
/*
 * 2008+ Copyright (c) Evgeniy Polyakov <zbr@ioremap.net>
 * 2011+ Copyright (c) Anton Kortunov <toshic.toshic@gmail.com>
 * All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

#include <errno.h>

#include <elliptics/packet.h>
#include <elliptics/interface.h>

#include "common.h"


int dnet_create_write_meta(struct dnet_meta_create_control *ctl, void **data)
{
	struct dnet_meta_container mc;
	struct dnet_meta_check_status *c;
	struct dnet_meta_checksum *csum;
	struct dnet_meta *m;
	int size = 0, err, nsize = 0;

	size += sizeof(struct dnet_meta_check_status) + sizeof(struct dnet_meta);

	if (ctl->obj && ctl->len)
		size += ctl->len + sizeof(struct dnet_meta);

	if (ctl->groups && ctl->group_num)
		size += ctl->group_num * sizeof(int) + sizeof(struct dnet_meta);

	size += sizeof(struct dnet_meta_checksum) + sizeof(struct dnet_meta);

	size += sizeof(struct dnet_meta_update) + sizeof(struct dnet_meta);

	if (!size) {
		err = -EINVAL;
		goto err_out_exit;
	}

	memset(&mc, 0, sizeof(struct dnet_meta_container));
	mc.data = malloc(size);
	if (!mc.data) {
		err = -ENOMEM;
		goto err_out_exit;
	}
	memset(mc.data, 0, size);

	m = (struct dnet_meta *)(mc.data);

	c = (struct dnet_meta_check_status *)m->data;
	m->size = sizeof(struct dnet_meta_check_status);
	m->type = DNET_META_CHECK_STATUS;

	/* Check status is undefined for now, it will be filled during actual check */
	memset(c, 0, sizeof(struct dnet_meta_check_status));
	dnet_convert_meta(m);

	m = (struct dnet_meta *)(m->data + m->size);
	dnet_create_meta_update(m, ctl->ts.tv_sec ? &ctl->ts : NULL, 0, 0);

	m = (struct dnet_meta *)(m->data + m->size);

	if (ctl->obj && ctl->len) {
		m->size = ctl->len;
		m->type = DNET_META_PARENT_OBJECT;
		memcpy(m->data, ctl->obj, ctl->len);
		dnet_convert_meta(m);

		m = (struct dnet_meta *)(m->data + m->size);
	}

	if (ctl->groups && ctl->group_num) {
		m->size = ctl->group_num * sizeof(int);
		m->type = DNET_META_GROUPS;
		memcpy(m->data, ctl->groups, ctl->group_num * sizeof(int));
		dnet_convert_meta(m);

		m = (struct dnet_meta *)(m->data + m->size);
	}

	csum = (struct dnet_meta_checksum *)m->data;
	memcpy(csum, ctl->checksum, DNET_CSUM_SIZE);
	csum->tm.tsec = ctl->ts.tv_sec;
	csum->tm.tnsec = ctl->ts.tv_nsec;
	dnet_convert_meta_checksum(csum);
	m->size = sizeof(struct dnet_meta_checksum);
	m->type = DNET_META_CHECKSUM;
	m = (struct dnet_meta *)(m->data + m->size);
	dnet_convert_meta(m);

	mc.size = size;
	memcpy(&mc.id, &ctl->id, sizeof(struct dnet_id));

	*data = mc.data;
	return mc.size;

err_out_exit:
	return err;
}

struct dnet_meta * dnet_meta_search_cust(struct dnet_meta_container *mc, uint32_t type)
{
	void *data = mc->data;
	uint32_t size = mc->size;
	struct dnet_meta m, *found = NULL;

	while (size) {
		if (size < sizeof(struct dnet_meta)) {
			fprintf(stderr, "Metadata size %u is too small, min %zu, searching for 0x%x.\n",
					size, sizeof(struct dnet_meta), type);
			break;
		}

		m = *(struct dnet_meta *)data;
		dnet_convert_meta(&m);

		if (m.size + sizeof(struct dnet_meta) > size) {
			fprintf(stderr, "Metadata entry broken: entry size %u, type: 0x%x, struct size: %zu, "
					"total size left: %u, searching for 0x%x.\n",
					m.size, m.type, sizeof(struct dnet_meta), size, type);
			break;
		}

		if (m.type == type) {
			found = data;
			break;
		}

		data += m.size + sizeof(struct dnet_meta);
		size -= m.size + sizeof(struct dnet_meta);
	}

	return found;
}

void dnet_common_log(void *priv __attribute((unused)), uint32_t mask, const char *msg)
{
        char str[64];
        struct tm tm;
        struct timeval tv;

        gettimeofday(&tv, NULL);
        localtime_r((time_t *)&tv.tv_sec, &tm);
        strftime(str, sizeof(str), "%F %R:%S", &tm);

        fprintf(stderr, "%s.%06lu %1x: %s", str, tv.tv_usec, mask, msg);
	fflush(stderr);
}

int dnet_parse_groups(char *value, int **groupsp)
{
	int len = strlen(value), i, num = 0, start = 0, pos = 0;
	char *ptr = value;
	int *groups;

	if (sscanf(value, "auto%d", &num) == 1) {
		*groupsp = NULL;
		return num;
	}

	for (i=0; i<len; ++i) {
		if (value[i] == DNET_CONF_ADDR_DELIM)
			start = 0;
		else if (!start) {
			start = 1;
			num++;
		}
	}

	if (!num) {
		fprintf(stderr, "no groups found\n");
		return -ENOENT;
	}

	groups = malloc(sizeof(int) * num);
	if (!groups)
		return -ENOMEM;

	memset(groups, 0, num * sizeof(int));

	start = 0;
	for (i=0; i<len; ++i) {
		if (value[i] == DNET_CONF_ADDR_DELIM) {
			value[i] = '\0';
			if (start) {
				groups[pos] = atoi(ptr);
				pos++;
				start = 0;
			}
		} else if (!start) {
			ptr = &value[i];
			start = 1;
		}
	}

	if (start) {
		groups[pos] = atoi(ptr);
		pos++;
	}

	*groupsp = groups;
	return pos;
}




common.h
/*
 * 2008+ Copyright (c) Evgeniy Polyakov <zbr@ioremap.net>
 * 2011+ Copyright (c) Anton Kortunov <toshic.toshic@gmail.com>
 * All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

#ifndef __COMMON_H
#define __COMMON_H

#include <time.h>

#include <elliptics/packet.h>
#include <elliptics/interface.h>

#ifdef __cplusplus
extern "C" {
#endif

#define DNET_CONF_ADDR_DELIM ':'

struct dnet_meta_create_control {
	struct dnet_id			id;
	const char			*obj;
	int				len;

	int				*groups;
	int				group_num;

	uint64_t			update_flags;
	struct timespec			ts;

	uint8_t				checksum[DNET_CSUM_SIZE];
};

int dnet_create_write_meta(struct dnet_meta_create_control *ctl, void **data);
struct dnet_meta * dnet_meta_search_cust(struct dnet_meta_container *mc, uint32_t type);
void dnet_common_log(void *priv __attribute((unused)), uint32_t mask, const char *msg);
int dnet_parse_groups(char *value, int **groupsp);

#ifdef __cplusplus
}
#endif

#endif /* __COMMON_H */


convert_meta.c
/*
 * 2008+ Copyright (c) Evgeniy Polyakov <zbr@ioremap.net>
 * 2011+ Copyright (c) Anton Kortunov <toshic.toshic@gmail.com>
 * All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <sys/mman.h>

#include <ctype.h>
#include <fcntl.h>
#include <errno.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <kclangc.h>

#include <elliptics/packet.h>
#include <elliptics/interface.h>
#include <eblob/blob.h>

#include "common.h"

static void mparser_usage(const char *p)
{
	fprintf(stderr, "Usage: %s args\n", p);
	fprintf(stderr, " -M                   - meta database to parse\n"
			" -N                   - new meta database (blob)\n"
			" -g                   - default groups for objects without groups in meta\n"
			" -h                   - this help\n");
	exit(-1);
}

uint64_t counter = 0;
uint64_t total = 0;

int *groups = NULL;
int group_num = 0;

struct db_ptrs {
	struct eblob_backend *newmeta;
};

static const char *mparser_visit(const char *key, size_t keysz,
			const char *mdata, size_t datasz, size_t *sp __attribute((unused)), void *opq)
{
	struct db_ptrs *ptrs = opq;
	char id_str[2 * DNET_ID_SIZE + 1];
	struct dnet_raw_id id;
	struct dnet_meta_container mc;
	struct dnet_meta_create_control ctl;
	struct dnet_meta m, *mp = NULL;
	struct dnet_meta_update *mu;
	struct dnet_meta_checksum *csum;
	void *data = (void *)mdata;
	unsigned int size = datasz;
	char tstr[64];
	time_t t;
	struct tm *tm;
	int err = 0;

	if (keysz != DNET_ID_SIZE) {
		fprintf(stdout, "Incorrect key size\n");
		goto err_out_exit;
	}

	memcpy(id.id, key, DNET_ID_SIZE);
	dnet_dump_id_len_raw(id.id, DNET_ID_SIZE, id_str);
	
	fprintf(stdout, "Processing key %.128s  ", id_str);

	memset(&ctl, 0, sizeof(ctl));
	dnet_setup_id(&ctl.id, 0, id.id);

	err = dnet_db_read_raw(ptrs->newmeta, &id, &mc.data);
	if (err != -ENOENT) {
		if (err > 0) {
			fprintf(stdout, "failed. Record with this ID already exists. Skipping.\n");
			goto err_out_free;
		} else {
			fprintf(stdout, "failed. Unable to read new meta, err %d\n", err);
			goto err_out_exit;
		}
	}

	while (size) {
		if (size < sizeof(struct dnet_meta)) {
			fprintf(stdout, "failed. Metadata size %u is too small, min %zu.\n",
					size, sizeof(struct dnet_meta));
			err = -1;
			goto err_out_exit;
			break;
		}

		mp = data;
		m = *(struct dnet_meta *)data;
		dnet_convert_meta(&m);

		if (m.size + sizeof(struct dnet_meta) > size) {
			fprintf(stdout , "failed. Metadata entry broken: entry size %u, type: 0x%x, struct size: %zu, "
					"total size left: %u.\n",
					m.size, m.type, sizeof(struct dnet_meta), size);
			err = -1;
			goto err_out_exit;
			break;
		}

		switch (m.type) {
			case DNET_META_PARENT_OBJECT:
				ctl.obj = mp->data;
				ctl.len = m.size;
				break;

			case DNET_META_GROUPS:
				ctl.groups = (int *)mp->data;
				ctl.group_num = m.size / sizeof(int);
				break;

			case DNET_META_CHECKSUM:
				csum = (struct dnet_meta_checksum *)mp->data;
				memcpy(ctl.checksum, csum->checksum, DNET_CSUM_SIZE);
				break;

			case DNET_META_UPDATE:
				mu = (struct dnet_meta_update *)mp->data;
				dnet_convert_meta_update(mu);
				ctl.ts.tv_sec = mu->tm.tsec;
				ctl.ts.tv_nsec = mu->tm.tnsec;
		}

		data += m.size + sizeof(struct dnet_meta);
		size -= m.size + sizeof(struct dnet_meta);
	}

	err = dnet_create_write_meta(&ctl, &mc.data);
	if (err <= 0) {
		fprintf(stdout, "failed to create new meta, err %d.\n", err);
		goto err_out_exit;
	}

	mc.size = err;

	err = dnet_db_write_raw(ptrs->newmeta, &id, mc.data, mc.size);
	if (err) {
		fprintf(stdout, "failed to write new meta, err %d.\n", err);
		goto err_out_free;
	}

	fprintf(stdout, "ok.\n");

err_out_free:
	free(mc.data);
err_out_exit:
	counter++;
	if (!(counter % 10000)) {
		t = time(NULL);
		tm = localtime(&t);
		strftime(tstr, sizeof(tstr), "%F %R:%S %Z", tm);
		fprintf(stderr, "%s: %llu/%llu records processed\n", tstr, counter, total);
	}

	return KCVISNOP;
}

int main(int argc, char *argv[])
{
	int err, ch;
	char *meta_name = NULL, *newmeta_name = NULL;
	unsigned long long offset, size;
	KCDB *meta = NULL;
	struct eblob_backend *newmeta = NULL;
	struct eblob_config ecfg;
	struct eblob_log log;
	char tstr[64];
	time_t t;
	struct tm *tm;
	struct db_ptrs ptrs;

	size = offset = 0;

	while ((ch = getopt(argc, argv, "M:N:g:h")) != -1) {
		switch (ch) {
			case 'M':
				meta_name = optarg;
				break;
			case 'N':
				newmeta_name = optarg;
				break;
			case 'g':
				group_num = dnet_parse_groups(optarg, &groups);
				break;
			case 'h':
				mparser_usage(argv[0]);
		}
	}

	if (!meta_name || !newmeta_name) {
		fprintf(stderr, "You have to provide meta database to convert.\n");
		mparser_usage(argv[0]);
	}

	if (group_num <= 0) {
		fprintf(stderr, "You have to provide default groups.\n");
		mparser_usage(argv[0]);
	}

	memset(&ptrs, 0, sizeof(struct db_ptrs));

	printf("opening %s meta database\n", meta_name);
	meta = kcdbnew();
	err = kcdbopen(meta, meta_name, KCOREADER | KCONOREPAIR);
	if (!err) {
		fprintf(stderr, "Failed to open meta database '%s': %d.\n", meta_name, -kcdbecode(meta));
		goto err_out_exit;
	}

	printf("opening %s new meta database\n", newmeta_name);

	memset(&ecfg, 0, sizeof(ecfg));
	ecfg.file = newmeta_name;
	ecfg.sync = 30;

	log.log = dnet_common_log;
	log.log_private = NULL;
	log.log_mask = EBLOB_LOG_ERROR | EBLOB_LOG_INFO | EBLOB_LOG_NOTICE;
	ecfg.log = &log;

	newmeta = eblob_init(&ecfg);
	if (!newmeta) {
		fprintf(stderr, "Failed to open meta database '%s'.\n", newmeta_name);
		goto err_out_dbopen;
	}

	ptrs.newmeta = newmeta;

	t = time(NULL);
	tm = localtime(&t);
	strftime(tstr, sizeof(tstr), "%F %R:%S %Z", tm);
	total = (unsigned long long)kcdbcount(meta);
	fprintf(stderr, "%s: Total %llu records in old meta DB\n", tstr, total);

	err = kcdbiterate(meta, mparser_visit, &ptrs, 0);
	if (!err) {
		fprintf(stderr, "Failed to iterate meta database '%s': %d.\n", meta_name, -kcdbecode(meta));
	}

	t = time(NULL);
	tm = localtime(&t);
	strftime(tstr, sizeof(tstr), "%F %R:%S %Z", tm);
	total = (unsigned long long)kcdbcount(meta);
	fprintf(stderr, "%s: Totally processed %llu records from history DB\n", tstr, counter);

err_out_dbopen2:
	eblob_cleanup(newmeta);

err_out_dbopen:
	err = kcdbclose(meta);
	if (!err)
		fprintf(stderr, "Failed to close meta database '%s': %d.\n", meta_name, -kcdbecode(meta));
	err = !err;
	kcdbdel(meta);

err_out_exit:
	return err;
}


blob_unsort.cpp

#include <sys/mman.h>
#include <sys/types.h>
#include <sys/stat.h>

#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <string.h>

#include <iostream>
#include <fstream>
#include <string>

#include <boost/iostreams/device/mapped_file.hpp>
#include <boost/program_options.hpp>
#include <boost/thread.hpp>
#include <boost/bind.hpp>
#include <boost/lexical_cast.hpp>
#include <boost/filesystem.hpp>
namespace fs = boost::filesystem;

#include <eblob/blob.h>

#include "common.h"


boost::iostreams::mapped_file file_;
std::map<uint64_t, struct eblob_disk_control *> blob_index;

void open_index(std::string &path) {
	std::ostringstream filename;
	struct eblob_disk_control *dc;
	uint64_t index_pos;


	file_.open(path, std::ios_base::in | std::ios_base::binary);

	index_pos = 0;
	while (index_pos < file_.size()) {

		dc = (struct eblob_disk_control *)(file_.const_data() + index_pos);

		blob_index[dc->position] = dc;

		index_pos += sizeof(struct eblob_disk_control);
	}
}

int main(int argc, char *argv[])
{
	std::map<uint64_t, struct eblob_disk_control *>::iterator iter;
	struct eblob_disk_control *dc;

	try {
		if (argc != 3) {
			std::ostringstream error_text;
			error_text << "Usage: " << argv[0] << " input_blob output_blob";

			throw std::runtime_error(error_text.str());
		}

		std::string input_path(argv[1]);

		std::ofstream unsorted_index(argv[2], std::ios::out | std::ios::binary);

		open_index(input_path);

		std::cout << "loaded " << blob_index.size() << " elements" << std::endl;

		for (iter = blob_index.begin(); iter != blob_index.end(); iter++) {
			dc = iter->second;
			unsorted_index.write((char *)dc, sizeof(struct eblob_disk_control));
		}

		unsorted_index.close();
		file_.close();

	} catch (const std::exception &e) {
		std::cerr << "Exiting: " << e.what() << std::endl;
		return -1;
	}
}


convert_files.cpp

#include <sys/mman.h>
#include <sys/types.h>
#include <sys/stat.h>

#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <string.h>

#include <iostream>
#include <string>

#include <boost/iostreams/device/mapped_file.hpp>
#include <boost/program_options.hpp>
#include <boost/thread.hpp>
#include <boost/bind.hpp>
#include <boost/lexical_cast.hpp>
#include <boost/filesystem.hpp>
#include <boost/date_time/posix_time/posix_time.hpp>
namespace fs = boost::filesystem;

#include <eblob/blob.h>

#include "common.h"

//using namespace zbr;

class processor_key {
	public:
		std::string path;
		uint64_t offset;
		uint64_t size;
		std::string id;
		boost::shared_ptr<boost::iostreams::mapped_file> file;
};

class generic_processor {
	public:
		virtual processor_key next(void) = 0;
};


struct timespec parse_time(std::string &datetime)
{
	struct timespec datetime_dt;
	try {
		boost::posix_time::ptime ptime(boost::posix_time::time_from_string(datetime));
		boost::posix_time::ptime epoch(boost::gregorian::date(1970,1,1));
		boost::posix_time::time_duration diff;

		diff = ptime - epoch;

		datetime_dt.tv_sec = diff.total_seconds();
		datetime_dt.tv_nsec = diff.total_nanoseconds();
	} catch(...) {
		datetime_dt.tv_sec = 0;
		datetime_dt.tv_nsec = 0;
	}

	return datetime_dt;
}

class eblob_processor : public generic_processor {
	public:
		eblob_processor(const std::string &path) : path_(path), index_(0) {
			open_index();
		}

		virtual ~eblob_processor() {
			if (file_.is_open())
				file_.close();

		}

		processor_key next(void) {
			struct eblob_disk_control dc;
			processor_key key;

			while (true) {
				if (pos_ >= file_.size()) {
					open_index();
				}

				memcpy(&dc, file_.const_data() + pos_, sizeof(dc));

				std::cout << "offset: " << dc.position << ", size: " << dc.data_size <<
					", disk_size: " << dc.disk_size << std::endl;

				pos_ += sizeof(dc);

				if (!(dc.flags & BLOB_DISK_CTL_REMOVE)) {
					key.id.assign((char *)dc.key.id, sizeof(dc.key.id));
					key.path = path_ + "." + boost::lexical_cast<std::string>(index_ - 1);
					key.offset = dc.position + sizeof(dc);
					key.size = dc.data_size;
					key.file = data_file_;
					break;
				}
			}
			return key;
		}

	private:
		std::string path_;
		int index_;
		uint64_t pos_;
		boost::iostreams::mapped_file file_;
		boost::shared_ptr<boost::iostreams::mapped_file> data_file_;

		void open_index() {
			std::ostringstream filename;
			struct eblob_disk_control *dc;
			uint64_t index_pos;

			pos_ = 0;

			if (file_.is_open())
				file_.close();

			filename << path_ << "." << index_;
			data_file_.reset(new boost::iostreams::mapped_file(filename.str(), std::ios_base::in | std::ios_base::binary));

			filename << ".index";
			file_.open(filename.str(), std::ios_base::in | std::ios_base::binary);

			++index_;

		}
};

class fs_processor : public generic_processor {
	public:
		fs_processor(const std::string &path) : itr_(fs::path(path)) {
		}

		processor_key next(void) {
			processor_key key;

			while (true) {
				if (itr_ == end_itr_) {
					throw std::runtime_error("Whole directory has been traversed");
				}

				if (fs::is_directory(itr_->path())) {
					++itr_;
					continue;
				}

				if (itr_->leaf().size() != DNET_ID_SIZE * 2) {
					++itr_;
					continue;
				}

				parse(itr_->leaf(), key.id);
				key.path = itr_->path().string();
				key.offset = 0;
				key.size = fs::file_size(itr_->path());

				if (key.size == 0) {
					++itr_;
					continue;
				}

				std::cout << "fs: " << itr_->path() << std::endl;
				key.file.reset(new boost::iostreams::mapped_file(key.path, std::ios_base::in | std::ios_base::binary));

				++itr_;
				break;
			}
			return key;
		}

	private:
		fs::recursive_directory_iterator end_itr_, itr_;
		std::vector<std::string> dirs_;

		void parse(const std::string &value, std::string &key) {
			unsigned char ch[5];
			unsigned int i, len = value.size();
			unsigned char id[DNET_ID_SIZE];

			memset(id, 0, DNET_ID_SIZE);

			if (len/2 > DNET_ID_SIZE)
				len = DNET_ID_SIZE * 2;

			ch[0] = '0';
			ch[1] = 'x';
			ch[4] = '\0';
			for (i=0; i<len / 2; i++) {
				ch[2] = value[2*i + 0];
				ch[3] = value[2*i + 1];

				id[i] = (unsigned char)strtol((const char *)ch, NULL, 16);
			}

			if (len & 1) {
				ch[2] = value[2*i + 0];
				ch[3] = '0';

				id[i] = (unsigned char)strtol((const char *)ch, NULL, 16);
			}

			key.assign((char *)id, DNET_ID_SIZE);
		}

};

class remote_update {
	public:
		remote_update(const std::vector<int> groups, const std::string meta, struct timespec update_date) :
				 groups_(groups), meta_(meta), update_date_(update_date), aflags_(0) {
		}

		void process(const std::string &path, int tnum = 16, int csum_enabled = 0) {
			generic_processor *proc;
			struct eblob_backend *meta = NULL;
			struct eblob_config ecfg;
			struct eblob_log log;

			if (!csum_enabled)
				aflags_ |= DNET_ATTR_NOCSUM;

			if (fs::is_directory(fs::path(path))) {
				proc = new fs_processor(path);
			} else {
				proc = new eblob_processor(path);
			}

			total_cnt = 0;

			memset(&ecfg, 0, sizeof(ecfg));
			ecfg.file = (char *)meta_.c_str();
			ecfg.sync = 30;

			log.log = dnet_common_log;
			log.log_private = NULL;
			log.log_mask = EBLOB_LOG_ERROR | EBLOB_LOG_INFO | EBLOB_LOG_NOTICE;
			ecfg.log = &log;

			meta = eblob_init(&ecfg);
			if (!meta) {
				std::cerr << "Failed to open meta database" << meta_ << std::endl;
				throw std::runtime_error("Failed to open meta database");
			}

			try {
				boost::thread_group threads;
				for (int i=0; i<tnum; ++i) {
					threads.create_thread(boost::bind(&remote_update::process_data, this, proc, meta));
				}

				threads.join_all();
			} catch (const std::exception &e) {
				std::cerr << "Finished processing " << path << " : " << e.what() << std::endl;
				eblob_cleanup(meta);
				delete proc;
				std::cerr << "Totally processed " << total_cnt << " records" << std::endl;
				throw e;
			}
			std::cerr << "1Totally processed " << total_cnt << " records" << std::endl;
			eblob_cleanup(meta);

			delete proc;
		}

	private:
		std::vector<int> groups_;
		std::string meta_;
		boost::mutex data_lock_;
		int aflags_;
		uint64_t total_cnt;
		struct timespec update_date_;

		void update(generic_processor *proc, processor_key &key, struct eblob_backend *meta) {
			struct dnet_raw_id id;
			struct dnet_meta *m;
			struct dnet_meta_container mc;
			struct dnet_meta_checksum *csum;
			struct dnet_meta *mp;
			uint8_t checksum[DNET_CSUM_SIZE];
			int err;

			if (key.offset + key.size > key.file->size()) {
				std::cout << "failed. " << dnet_dump_id_str(id.id) << ": incorrect length: "
				<< "offset=" << key.offset << ", size=" << key.size << ", file.size=" << key.file->size() << std::endl;
				return;
			}

			memset(&mc, 0, sizeof(mc));

			memcpy(&mc.id, (unsigned char *)key.id.data(), DNET_ID_SIZE);
			std::cout << "Processing " << dnet_dump_id_len(&mc.id, DNET_ID_SIZE) << " ";

			memcpy(&id.id, (unsigned char *)key.id.data(), DNET_ID_SIZE);
			err = dnet_db_read_raw(meta, &id, &mc.data);
			if (err == -ENOENT) {
				struct dnet_meta_create_control ctl;

				std::cout << "not found. Re-creating metadata" << std::endl;

				memset(&ctl, 0, sizeof(ctl));

				ctl.obj = NULL;
				ctl.len = 0;

				ctl.groups = &groups_[0];
				ctl.group_num = groups_.size();

				if (!(aflags_ & DNET_ATTR_NOCSUM)) {
					eblob_hash(meta, ctl.checksum, sizeof(ctl.checksum), key.file->const_data() + key.offset, key.size);
				}

				dnet_setup_id(&ctl.id, 0, id.id);

				ctl.ts = update_date_;

				err = dnet_create_write_meta(&ctl, &mc.data);
				if (err <= 0) {
					std::cout << "Metadata re-creating failed! err: " << err << std::endl;
					return;
				}

				mc.size = err;
				err = dnet_db_write_raw(meta, &id, mc.data, mc.size);
				if (err) {
					std::cout << "Metadata write failed! err: " << err << std::endl;
				}

			} else if (err <= 0) {
				std::cout << "failed. " << dnet_dump_id_str(id.id) << ": meta DB read failed, err: " << err << std::endl;
				return;
			} else if (err > 0 && !(aflags_ & DNET_ATTR_NOCSUM)) {
				mc.size = err;

				mp = dnet_meta_search_cust(&mc, DNET_META_CHECKSUM);
				if (mp) {
					csum = (struct dnet_meta_checksum *)mp->data;
					eblob_hash(meta, checksum, sizeof(checksum), key.file->const_data() + key.offset, key.size);
					if (memcmp(csum->checksum, checksum, DNET_CSUM_SIZE)) {
						std::cout << "Checksum mismatch, updating with the new one" << std::endl;

						memcpy(csum->checksum, checksum, DNET_CSUM_SIZE);
						dnet_current_time(&csum->tm);
						dnet_convert_meta_checksum(csum);

						err = dnet_db_write_raw(meta, &id, mc.data, mc.size);
						if (err) {
							std::cout << "Metadata write failed! err: " << err << std::endl;
						}
					}
				}

			}

			free(mc.data);
		}

		void process_data(generic_processor *proc, struct eblob_backend *meta) {
			try {
				while (true) {
					processor_key key;

					{
						boost::mutex::scoped_lock scoped_lock(data_lock_);
						key = proc->next();
						total_cnt++;
					}

					update(proc, key, meta);
				}
			} catch (const std::exception &e) {
				std::cerr << "Catched exception : " << e.what() << std::endl;
			}
		}
};

int main(int argc, char *argv[])
{
	try {
		namespace po = boost::program_options;
		po::options_description desc("Options (required options are marked with *");
		int groups_array[] = {1};
		std::vector<int> groups(groups_array, groups_array + ARRAY_SIZE(groups_array));
		std::string addr;
		std::string meta;
		std::string update_date;
		struct timespec update_dt;
		int port, family;
		int thread_num;
		int csum_enabled;

		desc.add_options()
			("help", "This help message")
			("input-path", po::value<std::string>(), "Input path (*)")
			("threads", po::value<int>(&thread_num)->default_value(16), "Number of threads to iterate over input data")
			("group", po::value<std::vector<int> >(&groups),
			 	"Group number which will host given object, can be used multiple times for several groups")
			("meta", po::value<std::string>(&meta), "Meta DB")
			("enable-checksum", po::value<int>(&csum_enabled)->default_value(0),
			 	"Set to 1 if you want to enable server generated checksums")
			("update-date", po::value<std::string>(&update_date)->default_value(""),
				"Update date for created meta in format like \"2011-08-22 21:42:00\"")
		;

		po::variables_map vm;
		po::store(po::parse_command_line(argc, argv, desc), vm);
		po::notify(vm);

		if (vm.count("help") || !vm.count("input-path") || !vm.count("meta")) {
			std::cout << desc << "\n";
			return -1;
		}

		update_dt = parse_time(update_date);
		remote_update up(groups, meta, update_dt);
		up.process(vm["input-path"].as<std::string>(), thread_num, csum_enabled);
	} catch (const std::exception &e) {
		std::cerr << "Exiting: " << e.what() << std::endl;
	}
}
